File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void dfs(int i, int j, vector<vector<bool>> & visited, vector<vector<char>>& grid)
    {
        int max_rows= grid.size();
        int max_columns = grid[0].size();
             //top
            if(i-1 >=0 && !visited[i-1][j] && grid[i-1][j] == '1')
            
            {
                visited[i-1][j] = true;
                dfs(i-1, j, visited, grid);
            }
            //left
            if(j-1 >=0 && !visited[i][j-1] && grid[i][j -1 ] == '1')
            {
                 visited[i][j-1] = true;
                dfs(i, j-1, visited, grid);
            }
            //right
            if(j+1 < max_columns && !visited[i][j+1] && grid[i][j+1] == '1')
            {
                visited[i][j+1] = true;
                dfs(i, j+1, visited, grid);
            }
            //down
            if(i+1 < max_rows && !visited[i+1][j] &&  grid[i+1][j] == '1')
            {
                visited[i+1][j] = true;
                dfs(i+1, j, visited, grid);
            }
        
    }
    int numIslands(vector<vector<char>>& grid) {
        vector<vector<bool>> visited;

        int num_comp=0;
        for(int i=0;i<grid.size();i++)
        {
            vector<bool> vec;
            visited.push_back(vec);
            for(int j=0;j<grid[0].size();j++)
                visited[i].push_back(false);
        }
        
        for(int i=0;i<grid.size();i++)
        {
            for(int j=0;j<grid[0].size();j++)
            {
                if(!visited[i][j] && grid[i][j] == '1')
                {
                    num_comp++;
                    dfs(i,j, visited, grid);
                }
            }
        }
        return num_comp;
    }
};

File: 200-number-of-islands/number-of-islands.java
class Solution {
    boolean[][] visited;
    int num_comps = 0;
    public int numIslands(char[][] grid) 
    {
        visited = new  boolean[grid.length][grid[0].length];
        for(int i=0;i<grid.length;i++)
        {
            for(int j=0;j<grid[0].length;j++)
                visited[i][j] = false;
        }

        for(int i=0;i<grid.length;i++)
        {
            for(int j=0;j<grid[0].length;j++)
            {
                    if(!visited[i][j] && grid[i][j] == '1')
                    {
                        visited[i][j] = true;
                        dfs(i,j, grid);
                        this.num_comps++;
                    }

            }
        }
        return this.num_comps;
        
    }
    void dfs(int i, int j, char[][] grid)
    {
         if(i+1 < grid.length && !visited[i+1][j] && grid[i+1][j] == '1')
        {
                        visited[i+1][j] = true;
                        dfs(i+1,j, grid);
                        
        }
        if(i-1 >= 0 && !visited[i-1][j] && grid[i-1][j] == '1')
        {
                        visited[i-1][j] = true;
                        dfs(i-1,j, grid);
                        
        }

         if(j+1 < grid[0].length && !visited[i][j+1] && grid[i][j+1] == '1')
        {
                        visited[i][j+1] = true;
                        dfs(i,j+1, grid);
                        
        }
        if(j-1 >= 0 && !visited[i][j-1] && grid[i][j-1] == '1')
        {
                        visited[i][j-1] = true;
                        dfs(i,j-1, grid);
                        
        }

    }

}